Extension will load failed, if the macOS Finder has opened ~/Library/Application\ Support/Zed/extensions/installed/ folder

### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

I have found may be there a bug with the extension file watching.

1. Install a language extension, for example: `nu`
2. Run script `open ~/Library/Application Support/Zed/extensions/installed/nu` to open extension folder in Finder.
3. Then the extension will fail, and then it will break.

This is may the extension watching bug, may when we open the Finder, may be the Finder will lock some file or folder, then Zed and Finder are conflict with this folder.

![20240221-095802](https://github.com/zed-industries/zed/assets/5518/276d618a-1109-45ff-8df9-ca6e2d4fdbd8)


### Environment

macOS

### If applicable, add mockups / screenshots to help explain present your vision of the feature

_No response_

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.
If you only need the most recent lines, you can run the `zed: open log` command palette action to see the last 1000.


```
2024-02-21T09:44:35+08:00 [INFO] 0 unhandled notification LogMessage:
{
  "level": 1,
  "message": "[INFO] [streamChoices] [2024-02-21T01:44:35.444Z] request done: headerRequestId: [d7ae317b-b140-4b1c-ad19-67d2b9677693] model deployment ID: [y5aa6d170bcae]",
  "metadataStr": "[INFO] [streamChoices] [2024-02-21T01:44:35.444Z]",
  "extra": [
    "request done: headerRequestId: [d7ae317b-b140-4b1c-ad19-67d2b9677693] model deployment ID: [y5aa6d170bcae]"
  ]
}
2024-02-21T09:44:57+08:00 [
ERROR] crates/extension/src/extension_store.rs:569: failed to load extension.json

Caused by:
    Not a directory (os error 20)
2024-02-21T09:44:57+08:00 [ERROR] crates/extension/src/extension_store.rs:569: Not a directory (os error 20)
2024-02-21T09:44:57+08:00 [ERROR] crates/extension/src/extension_store.rs:569: Not a directory (os error 20)
2024-02-21T09:44:57+08:00 [ERROR] crates/extension/src/extension_store.rs:569: failed to load extension.json

Caused by:
    No such file or directory (os error 2)
2024-02-21T09:44:57+08:00 [INFO] 0 unhandled notification LogMessage:
{
  "level": 1,
  "message": "[INFO] [ghostText] [2024-02-21T01:44:57.447Z] Found inline suggestions locally",
  "metadataStr": "[INFO] [ghostText] [2024-02-21T01:44:57.447Z]",
  "extra": [
    "Found inline suggestions locally"
  ]
}
2024-02-21T09:44:57+08:00 [INFO] 0 unhandled notification LogMessage:
{
  "level": 1,
  "message": "[INFO] [ghostText] [2024-02-21T01:44:57.512Z] Cancelled during debounce",
  "metadataStr": "[INFO] [ghostText] [2024-02-21T01:44:57.512Z]",
  "extra": [
    "Cancelled during debounce"
  ]
}
2024-02-21T09:45:01+08:00 [ERROR] crates/extension/src/extension_store.rs:569: failed to load extension.json

Caused by:
    Not a directory (os error 20)

2024-02-21T09:45:01+08:00 [ERROR] crates/extension/src/extension_store.rs:569: Not a directory (os error 20)
2024-02-21T09:45:01+08:00 [ERROR] crates/extension/src/extension_store.rs:569: Not a directory (os error 20)
2024-02-21T09:45:01+08:00 [ERROR] crates/extension/src/extension_store.rs:569: failed to load extension.json

Caused by:
    No such file or directory (os error 2)
2024-02-21T09:45:02+08:00 [INFO] 0 unhandled notification LogMessage:
{
  "level": 1,
  "message": "[INFO] [ghostText] [2024-02-21T01:45:02.123Z] Found inline suggestions locally",
  "metadataStr": "[INFO] [ghostText] [2024-02-21T01:45:02.123Z]",
  "extra": [
    "Found inline suggestions locally"
  ]
}
2024-02-21T09:45:02+08:00 [INFO] 0 unhandled notification LogMessage:
{
  "level": 1,
  "message": "[INFO] [ghostText] [2024-02-21T01:45:02.134Z] Cancelled during debounce",
  "metadataStr": "[INFO] [ghostText] [2024-02-21T01:45:02.134Z]",
  "extra": [
    "Cancelled during debounce"
  ]
}
2024-02-21T09:45:14+08:00 [INFO] 2 unhandled notification experimental/serverStatus:
{
  "health": "ok",
  "quiescent": true,
  "message": null
}
2024-02-21T09:45:23+08:00 [ERROR] crates/extension/src/extension_store.rs:569: failed to load extension.json

Caused by:
    Not a directory (os error 20)
2024-02-21T09:45:23+08:00 [ERROR] crates/extension/src/extension_store.rs:569: Not a directory (os error 20)
2024-02-21T09:45:23+08:00 [ERROR] crates/extension/src/extension_store.rs:569: Not a directory (os error 20)
2024-02-21T09:45:23+08:00 [ERROR] crates/extension/src/extension_store.rs:569: Not a directory (os error 20)
2024-02-21T09:45:23+08:00 [ERROR] crates/extension/src/extension_store.rs:569: failed to load extension.json

Caused by:
    No such file or directory (os error 2)
2024-02-21T09:45:23+08:00 [INFO] 0 unhandled notification LogMessage:
{
  "level": 1,
  "message": "[INFO] [ghostText] [2024-02-21T01:45:23.654Z] Found inline suggestions locally",
  "metadataStr": "[INFO] [ghostText] [2024-02-21T01:45:23.654Z]",
  "extra": [
    "Found inline suggestions locally"
  ]
}
2024-02-21T09:45:23+08:00 [INFO] 0 unhandled notification LogMessage:
{
  "level": 1,
  "message": "[INFO] [ghostText] [2024-02-21T01:45:23.705Z] Cancelled during debounce",
  "metadataStr": "[INFO] [ghostText] [2024-02-21T01:45:23.705Z]",
  "extra": [
    "Cancelled during debounce"
  ]
}
```

--- a/crates/extension/src/extension_store.rs
+++ b/crates/extension/src/extension_store.rs
@@ -651,6 +651,9 @@ impl ExtensionStore {
                 let Ok(relative_path) = language_path.strip_prefix(&extension_dir) else {
                     continue;
                 };
+                if !language_path.is_dir() {
+                    continue;
+                }
                 let config = fs.load(&language_path.join("config.toml")).await?;
                 let config = ::toml::from_str::<LanguageConfig>(&config)?;
 

--- a/crates/extension/src/extension_store.rs
+++ b/crates/extension/src/extension_store.rs
@@ -651,7 +651,10 @@ impl ExtensionStore {
                 let Ok(relative_path) = language_path.strip_prefix(&extension_dir) else {
                     continue;
                 };
-                if !language_path.is_dir() {
+                let Ok(Some(fs_metadata)) = fs.metadata(&language_path).await else {
+                    continue;
+                };
+                if !fs_metadata.is_dir {
                     continue;
                 }
                 let config = fs.load(&language_path.join("config.toml")).await?;

